<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>enrichment for nucleosome-free fragments and nucleosome signals — enrichedFragments • ATACseqQC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="enrichment for nucleosome-free fragments and nucleosome signals — enrichedFragments" />
<meta property="og:description" content="Get the enrichment signals for nucleosome-free fagments and
             nucleosomes." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ATACseqQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ATACseqQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>enrichment for nucleosome-free fragments and nucleosome signals</h1>
    
    <div class="hidden name"><code>enrichedFragments.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Get the enrichment signals for nucleosome-free fagments and
             nucleosomes.</p>
    </div>

    <pre class="usage"><span class='fu'>enrichedFragments</span><span class='op'>(</span>
  <span class='va'>bamfiles</span>,
  index <span class='op'>=</span> <span class='va'>bamfiles</span>,
  <span class='va'>gal</span>,
  <span class='va'>TSS</span>,
  <span class='va'>librarySize</span>,
  upstream <span class='op'>=</span> <span class='fl'>1010L</span>,
  downstream <span class='op'>=</span> <span class='fl'>1010L</span>,
  n.tile <span class='op'>=</span> <span class='fl'>101L</span>,
  normal.method <span class='op'>=</span> <span class='st'>"quantile"</span>,
  adjustFragmentLength <span class='op'>=</span> <span class='fl'>80L</span>,
  TSS.filter <span class='op'>=</span> <span class='fl'>0.5</span>,
  seqlev <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"chr"</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>22</span>, <span class='st'>"X"</span>, <span class='st'>"Y"</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bamfiles</th>
      <td><p>A vector of characters indicates the file names of bams.</p></td>
    </tr>
    <tr>
      <th>index</th>
      <td><p>The names of the index file of the 'BAM' file being processed;
This is given without the '.bai' extension.</p></td>
    </tr>
    <tr>
      <th>gal</th>
      <td><p>A GAlignmentsList object or a list of GAlignmentPairs.
If bamfiles is missing, gal is required.</p></td>
    </tr>
    <tr>
      <th>TSS</th>
      <td><p>an object of <a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a> indicates
the transcript start sites. All the width of TSS should equal to 1.
Otherwise, TSS will be reset to the center of input TSS.</p></td>
    </tr>
    <tr>
      <th>librarySize</th>
      <td><p>A vector of numeric indicates the library size. Output of
<a href='https://rdrr.io/pkg/ChIPpeakAnno/man/estLibSize.html'>estLibSize</a></p></td>
    </tr>
    <tr>
      <th>upstream, downstream</th>
      <td><p>numeric(1) or integer(1).
Upstream and downstream size from each TSS.</p></td>
    </tr>
    <tr>
      <th>n.tile</th>
      <td><p>numeric(1) or integer(1). The number of tiles to generate
for each element of TSS.</p></td>
    </tr>
    <tr>
      <th>normal.method</th>
      <td><p>character(1). Normalization methods,
could be "none" or "quantile".
See <a href='https://rdrr.io/pkg/limma/man/normalizebetweenarrays.html'>normalizeBetweenArrays</a>.</p></td>
    </tr>
    <tr>
      <th>adjustFragmentLength</th>
      <td><p>numeric(1) or integer(1).
The size of fragment to be adjusted to.
Default is set to half of the nucleosome size (80)</p></td>
    </tr>
    <tr>
      <th>TSS.filter</th>
      <td><p>numeric(1). The filter for signal strength of each TSS.
Default 0.5 indicates the average signal strength for the TSS
from upstream to downstream bins should be greater than 0.5.</p></td>
    </tr>
    <tr>
      <th>seqlev</th>
      <td><p>A vector of character indicates the sequence names to be
considered.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of matrixes. In each matrix, each row record the signals for
corresponding feature.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Jianhong Ou</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='va'>bamfiles</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"splited"</span>,
                       <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"NucleosomeFree.bam"</span>,
                         <span class='st'>"mononucleosome.bam"</span>,
                         <span class='st'>"dinucleosome.bam"</span>,
                         <span class='st'>"trinucleosome.bam"</span><span class='op'>)</span>, package<span class='op'>=</span><span class='st'>"ATACseqQC"</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class='op'>)</span>
<span class='va'>txs</span> <span class='op'>&lt;-</span> <span class='fu'>transcripts</span><span class='op'>(</span><span class='va'>TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class='op'>)</span>
<span class='va'>TSS</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html'>promoters</a></span><span class='op'>(</span><span class='va'>txs</span>, upstream<span class='op'>=</span><span class='fl'>0</span>, downstream<span class='op'>=</span><span class='fl'>1</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>ChIPpeakAnno</span><span class='op'>)</span>
<span class='va'>librarySize</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/ChIPpeakAnno/man/estLibSize.html'>estLibSize</a></span><span class='op'>(</span><span class='va'>bamfiles</span><span class='op'>)</span>
<span class='va'>sigs</span> <span class='op'>&lt;-</span> <span class='fu'>enrichedFragments</span><span class='op'>(</span><span class='va'>bamfiles</span>, TSS<span class='op'>=</span><span class='va'>TSS</span>, librarySize<span class='op'>=</span><span class='va'>librarySize</span>,
                          seqlev<span class='op'>=</span><span class='st'>"chr1"</span>, TSS.filter<span class='op'>=</span><span class='fl'>0</span><span class='op'>)</span>
<span class='va'>sigs.log2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='va'>sigs</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>.ele</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>.ele</span><span class='op'>+</span><span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedHeatmap.html'>featureAlignedHeatmap</a></span><span class='op'>(</span><span class='va'>sigs.log2</span>, <span class='fu'><a href='https://rdrr.io/pkg/ChIPpeakAnno/man/reCenterPeaks.html'>reCenterPeaks</a></span><span class='op'>(</span><span class='va'>TSS</span>, width<span class='op'>=</span><span class='fl'>2020</span><span class='op'>)</span>,
                      zeroAt<span class='op'>=</span><span class='fl'>.5</span>, n.tile<span class='op'>=</span><span class='fl'>101</span>, upper.extreme<span class='op'>=</span><span class='fl'>2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: GRanges object contains 1 out-of-bound range located on sequence</span>
#&gt; <span class='warning'>  chrUn_gl000223. Note that ranges located on a sequence whose length is</span>
#&gt; <span class='warning'>  unknown (NA) or on a circular sequence are not considered out-of-bound</span>
#&gt; <span class='warning'>  (use seqlengths() and isCircular() to get the lengths and circularity</span>
#&gt; <span class='warning'>  flags of the underlying sequences). You can use trim() to trim these</span>
#&gt; <span class='warning'>  ranges. See ?`trim,GenomicRanges-method` for more information.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: GRanges object contains 1 out-of-bound range located on sequence</span>
#&gt; <span class='warning'>  chrUn_gl000223. Note that ranges located on a sequence whose length is</span>
#&gt; <span class='warning'>  unknown (NA) or on a circular sequence are not considered out-of-bound</span>
#&gt; <span class='warning'>  (use seqlengths() and isCircular() to get the lengths and circularity</span>
#&gt; <span class='warning'>  flags of the underlying sequences). You can use trim() to trim these</span>
#&gt; <span class='warning'>  ranges. See ?`trim,GenomicRanges-method` for more information.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Some start position of the peaks are less than 1!</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Some end position of the peaks are out of bound!</span></div><div class='img'><img src='enrichedFragments-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'><a href='https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedDistribution.html'>featureAlignedDistribution</a></span><span class='op'>(</span><span class='va'>sigs</span>, <span class='fu'><a href='https://rdrr.io/pkg/ChIPpeakAnno/man/reCenterPeaks.html'>reCenterPeaks</a></span><span class='op'>(</span><span class='va'>TSS</span>, width<span class='op'>=</span><span class='fl'>2020</span><span class='op'>)</span>,
                           zeroAt<span class='op'>=</span><span class='fl'>.5</span>, n.tile<span class='op'>=</span><span class='fl'>101</span>, type<span class='op'>=</span><span class='st'>"l"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: GRanges object contains 1 out-of-bound range located on sequence</span>
#&gt; <span class='warning'>  chrUn_gl000223. Note that ranges located on a sequence whose length is</span>
#&gt; <span class='warning'>  unknown (NA) or on a circular sequence are not considered out-of-bound</span>
#&gt; <span class='warning'>  (use seqlengths() and isCircular() to get the lengths and circularity</span>
#&gt; <span class='warning'>  flags of the underlying sequences). You can use trim() to trim these</span>
#&gt; <span class='warning'>  ranges. See ?`trim,GenomicRanges-method` for more information.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: GRanges object contains 1 out-of-bound range located on sequence</span>
#&gt; <span class='warning'>  chrUn_gl000223. Note that ranges located on a sequence whose length is</span>
#&gt; <span class='warning'>  unknown (NA) or on a circular sequence are not considered out-of-bound</span>
#&gt; <span class='warning'>  (use seqlengths() and isCircular() to get the lengths and circularity</span>
#&gt; <span class='warning'>  flags of the underlying sequences). You can use trim() to trim these</span>
#&gt; <span class='warning'>  ranges. See ?`trim,GenomicRanges-method` for more information.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Some start position of the peaks are less than 1!</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Some end position of the peaks are out of bound!</span></div><div class='img'><img src='enrichedFragments-2.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


