<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>plot ATAC-seq footprints infer factor occupancy genome wide — factorFootprints • ATACseqQC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="plot ATAC-seq footprints infer factor occupancy genome wide — factorFootprints" />
<meta property="og:description" content="Aggregate ATAC-seq footprint for a given motif generated
             over binding sites within the genome." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ATACseqQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ATACseqQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>plot ATAC-seq footprints infer factor occupancy genome wide</h1>
    
    <div class="hidden name"><code>factorFootprints.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Aggregate ATAC-seq footprint for a given motif generated
             over binding sites within the genome.</p>
    </div>

    <pre class="usage"><span class='fu'>factorFootprints</span><span class='op'>(</span>
  <span class='va'>bamfiles</span>,
  index <span class='op'>=</span> <span class='va'>bamfiles</span>,
  <span class='va'>pfm</span>,
  <span class='va'>genome</span>,
  min.score <span class='op'>=</span> <span class='st'>"95%"</span>,
  <span class='va'>bindingSites</span>,
  seqlev <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"chr"</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>22</span>, <span class='st'>"X"</span>, <span class='st'>"Y"</span><span class='op'>)</span><span class='op'>)</span>,
  upstream <span class='op'>=</span> <span class='fl'>100</span>,
  downstream <span class='op'>=</span> <span class='fl'>100</span>,
  maxSiteNum <span class='op'>=</span> <span class='fl'>1e+06</span>,
  anchor <span class='op'>=</span> <span class='st'>"cut site"</span>,
  group <span class='op'>=</span> <span class='st'>"strand"</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bamfiles</th>
      <td><p>A vector of characters indicates the file names of bams.
All the bamfiles will be pulled together.</p></td>
    </tr>
    <tr>
      <th>index</th>
      <td><p>The names of the index file of the 'BAM' file being processed;
This is given without the '.bai' extension.</p></td>
    </tr>
    <tr>
      <th>pfm</th>
      <td><p>A Position frequency Matrix represented as a numeric matrix
with row names A, C, G and T.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>An object of <a href='https://rdrr.io/pkg/BSgenome/man/BSgenome-class.html'>BSgenome</a>.</p></td>
    </tr>
    <tr>
      <th>min.score</th>
      <td><p>The minimum score for counting a match.
Can be given as a character string containing a
percentage (e.g. "95<!-- %") of the highest possible -->
score or as a single number.
See <a href='https://rdrr.io/pkg/Biostrings/man/matchPWM.html'>matchPWM</a>.</p></td>
    </tr>
    <tr>
      <th>bindingSites</th>
      <td><p>A object of <a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a> indicates
candidate binding sites (eg. the output of fimo). The GRanges object must have
score column in the metadata column.</p></td>
    </tr>
    <tr>
      <th>seqlev</th>
      <td><p>A vector of characters indicates the sequence levels.</p></td>
    </tr>
    <tr>
      <th>upstream, downstream</th>
      <td><p>numeric(1) or integer(1).
Upstream and downstream of the binding region for
aggregate ATAC-seq footprint.</p></td>
    </tr>
    <tr>
      <th>maxSiteNum</th>
      <td><p>numeric(1). Maximal number of predicted binding sites.
if predicted binding sites is more than this number, top maxSiteNum binding
sites will be used.</p></td>
    </tr>
    <tr>
      <th>anchor</th>
      <td><p>"cut site" or "fragment center". If "fragment center" is used, 
the input bamfiles must be paired-end.</p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>Group information for the bamfiles. Default by strand. Otherwise,
a factor or vector of characters with same length of bamfiles. The
group is limited to 2 groups.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>xlab, legTitle, xlim or ylim could be used by <a href='plotFootprints.html'>plotFootprints</a></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>an invisible list of matrixes with the signals for plot.
It includes:
 - signal    mean values of coverage for positive strand and negative strand
             in feature regions
 - spearman.correlation    spearman correlations of cleavage counts in the 
             highest 10-nucleotide-window and binding prediction score.
 - bindingSites    predicted binding sites.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Chen, K., Xi, Y., Pan, X., Li, Z., Kaestner, K., Tyler, J.,
Dent, S., He, X. and Li, W., 2013.
DANPOS: dynamic analysis of nucleosome position and occupancy by sequencing.
Genome research, 23(2), pp.341-351.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Jianhong Ou, Julie Zhu</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='va'>bamfile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"GL1.bam"</span>,
                       package<span class='op'>=</span><span class='st'>"ATACseqQC"</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>MotifDb</span><span class='op'>)</span>
<span class='va'>CTCF</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/MotifDb/man/query.html'>query</a></span><span class='op'>(</span><span class='va'>MotifDb</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CTCF"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>CTCF</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>as.list</a></span><span class='op'>(</span><span class='va'>CTCF</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>BSgenome.Hsapiens.UCSC.hg19</span><span class='op'>)</span>
<span class='fu'>factorFootprints</span><span class='op'>(</span><span class='va'>bamfile</span>, pfm<span class='op'>=</span><span class='va'>CTCF</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>,
                genome<span class='op'>=</span><span class='va'>Hsapiens</span>,
                min.score<span class='op'>=</span><span class='st'>"95%"</span>, seqlev<span class='op'>=</span><span class='st'>"chr1"</span>,
                upstream<span class='op'>=</span><span class='fl'>100</span>, downstream<span class='op'>=</span><span class='fl'>100</span><span class='op'>)</span>
</div><div class='img'><img src='factorFootprints-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'>##### Using user defined binding sites #####</span>
<span class='va'>bds</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"jolma2013.motifs.bindingList.95.rds"</span>,
                          package<span class='op'>=</span><span class='st'>"ATACseqQC"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>bindingSites</span> <span class='op'>&lt;-</span> <span class='va'>bds</span><span class='op'>[[</span><span class='st'>"Hsapiens-jolma2013-CTCF"</span><span class='op'>]</span><span class='op'>]</span>
<span class='va'>seqlev</span> <span class='op'>&lt;-</span> <span class='st'>"chr1"</span> <span class='co'>#seqlevels(bindingSites)</span>
<span class='va'>bindingSites</span> <span class='op'>&lt;-</span> <span class='va'>bindingSites</span><span class='op'>[</span><span class='fu'>seqnames</span><span class='op'>(</span><span class='va'>bindingSites</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='va'>seqlev</span><span class='op'>]</span>
<span class='fu'>seqlevels</span><span class='op'>(</span><span class='va'>bindingSites</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>seqlev</span>
<span class='fu'>seqinfo</span><span class='op'>(</span><span class='va'>bindingSites</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>seqinfo</span><span class='op'>(</span><span class='va'>Hsapiens</span><span class='op'>)</span><span class='op'>[</span><span class='va'>seqlev</span><span class='op'>]</span>
<span class='fu'>factorFootprints</span><span class='op'>(</span><span class='va'>bamfile</span>, pfm<span class='op'>=</span><span class='va'>CTCF</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>,
                genome<span class='op'>=</span><span class='va'>Hsapiens</span>, seqlev<span class='op'>=</span><span class='va'>seqlev</span>,
                upstream<span class='op'>=</span><span class='fl'>100</span>, downstream<span class='op'>=</span><span class='fl'>100</span><span class='op'>)</span>
</div><div class='img'><img src='factorFootprints-2.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


