<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Nucleosome Free Regions (NFR) score — NFRscore • ATACseqQC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Nucleosome Free Regions (NFR) score — NFRscore" />
<meta property="og:description" content="NFR score is a raio between cut signal adjacent to TSS and that flanking 
the corresponding TSS. Each TSS window of 400 bp is first seperated into 3 sub-regions: 
the most upstream 150 bp (n1), the most downstream of 150 bp (n2), and the middle 100 bp (nf). 
Then the number of fragments with 5' ends overlapping each region are calculated for each TSS.
The NFR score for each TSS is calculated as NFR-score = log2(nf) - log2((n1+n2)/2). 
A plot can be generate with the NFR scores as Y-axis and the average signals of 400 bp window as X-axis,
very like a MA plot for gene expression data." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ATACseqQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ATACseqQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Nucleosome Free Regions (NFR) score</h1>
    
    <div class="hidden name"><code>NFRscore.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>NFR score is a raio between cut signal adjacent to TSS and that flanking 
the corresponding TSS. Each TSS window of 400 bp is first seperated into 3 sub-regions: 
the most upstream 150 bp (n1), the most downstream of 150 bp (n2), and the middle 100 bp (nf). 
Then the number of fragments with 5' ends overlapping each region are calculated for each TSS.
The NFR score for each TSS is calculated as NFR-score = log2(nf) - log2((n1+n2)/2). 
A plot can be generate with the NFR scores as Y-axis and the average signals of 400 bp window as X-axis,
very like a MA plot for gene expression data.</p>
    </div>

    <pre class="usage"><span class='fu'>NFRscore</span><span class='op'>(</span>
  <span class='va'>obj</span>,
  <span class='va'>txs</span>,
  seqlev <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/sets.html'>intersect</a></span><span class='op'>(</span><span class='fu'>seqlevels</span><span class='op'>(</span><span class='va'>obj</span><span class='op'>)</span>, <span class='fu'>seqlevels</span><span class='op'>(</span><span class='va'>txs</span><span class='op'>)</span><span class='op'>)</span>,
  nucleosomeSize <span class='op'>=</span> <span class='fl'>150</span>,
  nucleosomeFreeSize <span class='op'>=</span> <span class='fl'>100</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>obj</th>
      <td><p>an object of <a href='https://rdrr.io/pkg/GenomicAlignments/man/GAlignments-class.html'>GAlignments</a></p></td>
    </tr>
    <tr>
      <th>txs</th>
      <td><p>GRanges of transcripts</p></td>
    </tr>
    <tr>
      <th>seqlev</th>
      <td><p>A vector of characters indicates the sequence levels.</p></td>
    </tr>
    <tr>
      <th>nucleosomeSize</th>
      <td><p>numeric(1) or integer(1). Default is 150</p></td>
    </tr>
    <tr>
      <th>nucleosomeFreeSize</th>
      <td><p>numeric(1) or integer(1). Default is 100</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A object of <a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a> with NFR scores</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Jianhong Ou</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'> 
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://bioconductor.org/packages/GenomicRanges'>GenomicRanges</a></span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomeInfoDb</span></div><div class='input'><span class='va'>bamfile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"GL1.bam"</span>, 
                       package<span class='op'>=</span><span class='st'>"ATACseqQC"</span>, mustWork<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>gal1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='readBamFile.html'>readBamFile</a></span><span class='op'>(</span>bamFile<span class='op'>=</span><span class='va'>bamfile</span>, tag<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/character.html'>character</a></span><span class='op'>(</span><span class='fl'>0</span><span class='op'>)</span>, 
                    which<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></span><span class='op'>(</span><span class='st'>"chr1"</span>, <span class='fu'>IRanges</span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1e6</span><span class='op'>)</span><span class='op'>)</span>, 
                    asMates<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicFeatures</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: AnnotationDbi</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biobase</span></div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    'browseVignettes()'. To cite Bioconductor, see</span>
#&gt; <span class='message'>    'citation("Biobase")', and for packages 'citation("pkgname")'.</span></div><div class='input'><span class='va'>txs</span> <span class='op'>&lt;-</span> <span class='fu'>transcripts</span><span class='op'>(</span><span class='va'>TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class='op'>)</span>
<span class='va'>nfr</span> <span class='op'>&lt;-</span> <span class='fu'>NFRscore</span><span class='op'>(</span><span class='va'>gal1</span>, <span class='va'>txs</span><span class='op'>)</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


