<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>read in bam files — readBamFile • ATACseqQC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="read in bam files — readBamFile" />
<meta property="og:description" content="wraper for readGAlignments/readGAlignmentsList to read in bam files." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ATACseqQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ATACseqQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>read in bam files</h1>
    
    <div class="hidden name"><code>readBamFile.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>wraper for readGAlignments/readGAlignmentsList to read in bam files.</p>
    </div>

    <pre class="usage"><span class='fu'>readBamFile</span><span class='op'>(</span>
  <span class='va'>bamFile</span>,
  <span class='va'>which</span>,
  tag <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/character.html'>character</a></span><span class='op'>(</span><span class='fl'>0</span><span class='op'>)</span>,
  what <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"qname"</span>, <span class='st'>"flag"</span>, <span class='st'>"mapq"</span>, <span class='st'>"isize"</span>, <span class='st'>"seq"</span>, <span class='st'>"qual"</span>, <span class='st'>"mrnm"</span><span class='op'>)</span>,
  flag <span class='op'>=</span> <span class='fu'>scanBamFlag</span><span class='op'>(</span>isSecondaryAlignment <span class='op'>=</span> <span class='cn'>FALSE</span>, isUnmappedQuery <span class='op'>=</span> <span class='cn'>FALSE</span>,
    isNotPassingQualityControls <span class='op'>=</span> <span class='cn'>FALSE</span>, isSupplementaryAlignment <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>,
  asMates <span class='op'>=</span> <span class='cn'>FALSE</span>,
  bigFile <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bamFile</th>
      <td><p>character(1). Bam file name.</p></td>
    </tr>
    <tr>
      <th>which</th>
      <td><p>A <a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a>, <a href='https://rdrr.io/pkg/IRanges/man/IntegerRangesList-class.html'>IntegerRangesList</a>, 
or any object that can be coerced to a RangesList, or missing object, 
from which a IRangesList instance will be constructed. 
See <a href='https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html'>ScanBamParam</a>.</p></td>
    </tr>
    <tr>
      <th>tag</th>
      <td><p>A vector of characters indicates the tag names to be read.
See <a href='https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html'>ScanBamParam</a>.</p></td>
    </tr>
    <tr>
      <th>what</th>
      <td><p>A character vector naming the fields to return. 
Fields are described on the Rsamtools[scanBam] help page.</p></td>
    </tr>
    <tr>
      <th>flag</th>
      <td><p>An integer(2) vector used to filter reads based on their 
'flag' entry.</p></td>
    </tr>
    <tr>
      <th>asMates</th>
      <td><p>logical(1). Paired ends or not</p></td>
    </tr>
    <tr>
      <th>bigFile</th>
      <td><p>If the file take too much memory, set it to true to avoid read the reads into memory.
<a href='https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html'>scanBamFlag</a> helper function.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>parameters used by <a href='https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html'>readGAlignmentsList</a> 
or <a href='https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html'>readGAlignments</a></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A GAlignmentsList object when asMates=TRUE,
otherwise A GAlignments object.
If bigFile is set to TRUE, no reads will be read into memory at this step and 
empty GAlignments/GAlignmentsList will be returned.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Jianhong Ou</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>BSgenome.Hsapiens.UCSC.hg19</span><span class='op'>)</span>
<span class='va'>which</span> <span class='op'>&lt;-</span> <span class='fu'>as</span><span class='op'>(</span><span class='fu'>seqinfo</span><span class='op'>(</span><span class='va'>Hsapiens</span><span class='op'>)</span><span class='op'>[</span><span class='st'>"chr1"</span><span class='op'>]</span>, <span class='st'>"GRanges"</span><span class='op'>)</span>
<span class='va'>bamfile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"GL1.bam"</span>, 
                       package<span class='op'>=</span><span class='st'>"ATACseqQC"</span>, mustWork<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'>readBamFile</span><span class='op'>(</span><span class='va'>bamfile</span>, which<span class='op'>=</span><span class='va'>which</span>, asMates<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GAlignmentsList object of length 44357:
#&gt; [[1]]
#&gt; GAlignments object with 2 alignments and 7 metadata columns:
#&gt;       seqnames strand       cigar    qwidth     start       end     width
#&gt;          &lt;Rle&gt;  &lt;Rle&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;   [1]     chr1      +         47M        47    567414    567460        47
#&gt;   [2]     chr1      -         50M        50    567544    567593        50
#&gt;           njunc |       qname      flag      mapq     isize
#&gt;       &lt;integer&gt; | &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;   [1]         0 |         100        99        30       180
#&gt;   [2]         0 |         100       147        30      -180
#&gt;                           seq                    qual     mrnm
#&gt;                &lt;DNAStringSet&gt;          &lt;PhredQuality&gt; &lt;factor&gt;
#&gt;   [1] ATTTAGCTGA...AAATGATCTG CCCFFFFFHH...JJJJJJJJJJ     chr1
#&gt;   [2] GACATCGTAC...TCCACTATGT JJJJJJJHHJ...HHFFFFFCCC     chr1
#&gt;   -------
#&gt;   seqinfo: 25 sequences from an unspecified genome
#&gt; 
#&gt; [[2]]
#&gt; GAlignments object with 2 alignments and 7 metadata columns:
#&gt;       seqnames strand       cigar    qwidth     start       end     width
#&gt;          &lt;Rle&gt;  &lt;Rle&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;   [1]     chr1      -         50M        50 101701613 101701662        50
#&gt;   [2]     chr1      +         50M        50 101701398 101701447        50
#&gt;           njunc |       qname      flag      mapq     isize
#&gt;       &lt;integer&gt; | &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;   [1]         0 |      100003        83        42      -265
#&gt;   [2]         0 |      100003       163        42       265
#&gt;                           seq                    qual     mrnm
#&gt;                &lt;DNAStringSet&gt;          &lt;PhredQuality&gt; &lt;factor&gt;
#&gt;   [1] TGAATCTCTC...ACCTCTCCAT FFFFFFGHHH...FHFDEDFCB@     chr1
#&gt;   [2] GACCAAGTCC...CTAAGCAAAA @CCFFFFEHG...JJIGJIJJIJ     chr1
#&gt;   -------
#&gt;   seqinfo: 25 sequences from an unspecified genome
#&gt; 
#&gt; [[3]]
#&gt; GAlignments object with 2 alignments and 7 metadata columns:
#&gt;       seqnames strand       cigar    qwidth     start       end     width
#&gt;          &lt;Rle&gt;  &lt;Rle&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;   [1]     chr1      +         50M        50 101701631 101701680        50
#&gt;   [2]     chr1      -         50M        50 101701766 101701815        50
#&gt;           njunc |       qname      flag      mapq     isize
#&gt;       &lt;integer&gt; | &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;   [1]         0 |      100005        99        42       185
#&gt;   [2]         0 |      100005       147        42      -185
#&gt;                           seq                    qual     mrnm
#&gt;                &lt;DNAStringSet&gt;          &lt;PhredQuality&gt; &lt;factor&gt;
#&gt;   [1] GCAGCAGCCC...ATCTGGGACA CC@FFFFFHH...JIJJIJJIJE     chr1
#&gt;   [2] AAAGCCACTG...GGATCTGGGG DHHIIIIJIH...HHFFFFFCCC     chr1
#&gt;   -------
#&gt;   seqinfo: 25 sequences from an unspecified genome
#&gt; 
#&gt; ...
#&gt; &lt;44354 more elements&gt;</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


