<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>split bams into nucleosome free, mononucleosome,
dinucleosome and trinucleosome — splitGAlignmentsByCut • ATACseqQC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="split bams into nucleosome free, mononucleosome,
dinucleosome and trinucleosome — splitGAlignmentsByCut" />
<meta property="og:description" content="use random forest to split the reads into nucleosome free,
             mononucleosome, dinucleosome and trinucleosome. 
             The features used in random forest including 
             fragment length, GC content, and 
             UCSC phastCons conservation scores." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ATACseqQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ATACseqQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>split bams into nucleosome free, mononucleosome,
dinucleosome and trinucleosome</h1>
    
    <div class="hidden name"><code>splitGAlignmentsByCut.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>use random forest to split the reads into nucleosome free,
             mononucleosome, dinucleosome and trinucleosome. 
             The features used in random forest including 
             fragment length, GC content, and 
             UCSC phastCons conservation scores.</p>
    </div>

    <pre class="usage"><span class='fu'>splitGAlignmentsByCut</span><span class='op'>(</span>
  <span class='va'>obj</span>,
  <span class='va'>txs</span>,
  <span class='va'>genome</span>,
  <span class='va'>conservation</span>,
  <span class='va'>outPath</span>,
  breaks <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>100</span>, <span class='fl'>180</span>, <span class='fl'>247</span>, <span class='fl'>315</span>, <span class='fl'>473</span>, <span class='fl'>558</span>, <span class='fl'>615</span>, <span class='cn'>Inf</span><span class='op'>)</span>,
  labels <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"NucleosomeFree"</span>, <span class='st'>"inter1"</span>, <span class='st'>"mononucleosome"</span>, <span class='st'>"inter2"</span>, <span class='st'>"dinucleosome"</span>,
    <span class='st'>"inter3"</span>, <span class='st'>"trinucleosome"</span>, <span class='st'>"others"</span><span class='op'>)</span>,
  labelsOfNucleosomeFree <span class='op'>=</span> <span class='st'>"NucleosomeFree"</span>,
  labelsOfMononucleosome <span class='op'>=</span> <span class='st'>"mononucleosome"</span>,
  trainningSetPercentage <span class='op'>=</span> <span class='fl'>0.15</span>,
  cutoff <span class='op'>=</span> <span class='fl'>0.8</span>,
  halfSizeOfNucleosome <span class='op'>=</span> <span class='fl'>80L</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>obj</th>
      <td><p>an object of <a href='https://rdrr.io/pkg/GenomicAlignments/man/GAlignments-class.html'>GAlignments</a></p></td>
    </tr>
    <tr>
      <th>txs</th>
      <td><p>GRanges of transcripts</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>an object of BSgenome</p></td>
    </tr>
    <tr>
      <th>conservation</th>
      <td><p>an object of <a href='https://rdrr.io/pkg/GenomicScores/man/GScores-class.html'>GScores</a>.</p></td>
    </tr>
    <tr>
      <th>outPath</th>
      <td><p>folder to save the splitted alignments. If outPath is setting, 
the return of the function will not contain seq and qual fields.</p></td>
    </tr>
    <tr>
      <th>breaks</th>
      <td><p>a numeric vector for fragment size of nucleosome free,
mononucleosome, dinucleosome and trinucleosome. The breaks pre-defined
here is following the description of Greenleaf's paper (see reference).</p></td>
    </tr>
    <tr>
      <th>labels</th>
      <td><p>a character vector for labels of the levels 
of the resulting category.</p></td>
    </tr>
    <tr>
      <th>labelsOfNucleosomeFree, labelsOfMononucleosome</th>
      <td><p>character(1). The label
for nucleosome free and mononucleosome.</p></td>
    </tr>
    <tr>
      <th>trainningSetPercentage</th>
      <td><p>numeric(1) between 0 and 1. Percentage of 
trainning set from top coverage.</p></td>
    </tr>
    <tr>
      <th>cutoff</th>
      <td><p>numeric(1) between 0 and 1. cutoff value for prediction.</p></td>
    </tr>
    <tr>
      <th>halfSizeOfNucleosome</th>
      <td><p>numeric(1) or integer(1). Thre read length will
be adjusted to half of the nucleosome size to enhance the signal-to-noise
ratio.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a list of GAlignments</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Buenrostro, J.D., Giresi, P.G., Zaba, L.C., Chang, H.Y. and 
Greenleaf, W.J., 2013. Transposition of native chromatin for fast and 
sensitive epigenomic profiling of open chromatin, DNA-binding proteins and 
nucleosome position. Nature methods, 10(12), pp.1213-1218.</p>
<p>Chen, K., Xi, Y., Pan, X., Li, Z., Kaestner, K., Tyler, J., Dent, S., 
He, X. and Li, W., 2013. DANPOS: dynamic analysis of nucleosome position 
and occupancy by sequencing. Genome research, 23(2), pp.341-351.</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Jianhong Ou</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://bioconductor.org/packages/GenomicRanges'>GenomicRanges</a></span><span class='op'>)</span>
<span class='va'>bamfile</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"GL1.bam"</span>, 
                       package<span class='op'>=</span><span class='st'>"ATACseqQC"</span>, mustWork<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>tags</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"AS"</span>, <span class='st'>"XN"</span>, <span class='st'>"XM"</span>, <span class='st'>"XO"</span>, <span class='st'>"XG"</span>, <span class='st'>"NM"</span>, <span class='st'>"MD"</span>, <span class='st'>"YS"</span>, <span class='st'>"YT"</span><span class='op'>)</span>
<span class='va'>gal1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='readBamFile.html'>readBamFile</a></span><span class='op'>(</span>bamFile<span class='op'>=</span><span class='va'>bamfile</span>, tag<span class='op'>=</span><span class='va'>tags</span>, 
                    which<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></span><span class='op'>(</span><span class='st'>"chr1"</span>, <span class='fu'>IRanges</span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1e6</span><span class='op'>)</span><span class='op'>)</span>, 
                    asMates<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>gal1</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>mcols</span><span class='op'>(</span><span class='va'>gal1</span><span class='op'>)</span><span class='op'>$</span><span class='va'>qname</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>BSgenome.Hsapiens.UCSC.hg19</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class='op'>)</span>
<span class='va'>txs</span> <span class='op'>&lt;-</span> <span class='fu'>transcripts</span><span class='op'>(</span><span class='va'>TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>phastCons100way.UCSC.hg19</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicScores</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'GenomicScores'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:utils':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    citation</span></div><div class='input'><span class='fu'>splitGAlignmentsByCut</span><span class='op'>(</span><span class='va'>gal1</span>, txs<span class='op'>=</span><span class='va'>txs</span>, genome<span class='op'>=</span><span class='va'>Hsapiens</span>, 
                      conservation<span class='op'>=</span><span class='va'>phastCons100way.UCSC.hg19</span><span class='op'>)</span>
</div><div class='output co'>#&gt; GAlignmentsList object of length 8:
#&gt; $NucleosomeFree
#&gt; GAlignments object with 1196 alignments and 16 metadata columns:
#&gt;        seqnames strand       cigar    qwidth     start       end     width
#&gt;           &lt;Rle&gt;  &lt;Rle&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;     14     chr1      +         47M        47    565854    565900        47
#&gt;     14     chr1      -         47M        47    565854    565900        47
#&gt;     17     chr1      +         40M        40    565903    565942        40
#&gt;     17     chr1      -         40M        40    565903    565942        40
#&gt;     56     chr1      +         50M        50    566898    566947        50
#&gt;    ...      ...    ...         ...       ...       ...       ...       ...
#&gt;   5083     chr1      +         50M        50    995025    995074        50
#&gt;   5085     chr1      -         50M        50    995029    995078        50
#&gt;   5083     chr1      -         50M        50    995042    995091        50
#&gt;   5103     chr1      +         50M        50    995094    995143        50
#&gt;   5103     chr1      -         50M        50    995097    995146        50
#&gt;            njunc |       qname      flag      mapq     isize
#&gt;        &lt;integer&gt; | &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
#&gt;     14         0 |          14        99        31       -47
#&gt;     14         0 |          14       147        31       -47
#&gt;     17         0 |          17        99        31       -40
#&gt;     17         0 |          17       147        31       -40
#&gt;     56         0 |          56        99        42        93
#&gt;    ...       ... .         ...       ...       ...       ...
#&gt;   5083         0 |        5083        99        42        67
#&gt;   5085         0 |        5085        83        42       -62
#&gt;   5083         0 |        5083       147        42       -67
#&gt;   5103         0 |        5103       163        42        53
#&gt;   5103         0 |        5103        83        42       -53
#&gt;                            seq                    qual     mrnm        AS
#&gt;                 &lt;DNAStringSet&gt;          &lt;PhredQuality&gt; &lt;factor&gt; &lt;integer&gt;
#&gt;     14 CCCACCATCA...ACTTCTACCT CCCFFFFFHH...JJJJJJJJIJ     chr1         0
#&gt;     14 CCCACCATCA...ACTTCTACCT JJIHIJJJJJ...HHFFFDDCCB     chr1         0
#&gt;     17 GCCTAATCTA...TCCCTATATC CCCFFFFFHH...JJJJGJJIIJ     chr1         0
#&gt;     17 GCCTAATCTA...TCCCTATATC IJJJJJIIHE...FHFD?FDBBB     chr1         0
#&gt;     56 TCCTTACACC...TCAATTTCAT CCCFFFFFHH...JJJIIIJJJJ     chr1         0
#&gt;    ...                     ...                     ...      ...       ...
#&gt;   5083 GCGTGGCCCG...CGGAGCCCAG @@@DDDFFHG...DBDBB5@DBB     chr1         0
#&gt;   5085 GGCCCGGGGC...GCCCAGGCAG DDDBDDDDFF...HHFFFFFCCB     chr1         0
#&gt;   5083 GCATCGGGGC...CTTCGCGGAG ?BDDBBB?6&gt;...HHFDDFF@B@     chr1         0
#&gt;   5103 GCCGGAGGGT...GCGAGCGGCG ??BDDDDFFF...##########     chr1         0
#&gt;   5103 GGAGGGTGCG...AGCGGCGCGG &lt;@8@8575&amp;B...:?8@8BD=??     chr1        -4
#&gt;               XN        XM        XO        XG        NM          MD        YS
#&gt;        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
#&gt;     14         0         0         0         0         0          47         0
#&gt;     14         0         0         0         0         0          47         0
#&gt;     17         0         0         0         0         0          40         0
#&gt;     17         0         0         0         0         0          40         0
#&gt;     56         0         0         0         0         0          50        -3
#&gt;    ...       ...       ...       ...       ...       ...         ...       ...
#&gt;   5083         0         0         0         0         0          50         0
#&gt;   5085         0         0         0         0         0          50         0
#&gt;   5083         0         0         0         0         0          50         0
#&gt;   5103         0         0         0         0         0          50        -4
#&gt;   5103         0         2         0         0         2     24A7C17         0
#&gt;                 YT
#&gt;        &lt;character&gt;
#&gt;     14          CP
#&gt;     14          CP
#&gt;     17          CP
#&gt;     17          CP
#&gt;     56          CP
#&gt;    ...         ...
#&gt;   5083          CP
#&gt;   5085          CP
#&gt;   5083          CP
#&gt;   5103          CP
#&gt;   5103          CP
#&gt;   -------
#&gt;   seqinfo: 25 sequences from an unspecified genome
#&gt; 
#&gt; ...
#&gt; &lt;7 more elements&gt;</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


